### PR TITLE
tomcat: use systemd where available due to newer versions not generating PID-file

### PR DESCRIPTION
**Pull request by oalbrigt**

### CODE REVIEW

**Comment by oalbrigt** [comment]
Newer Tomcat version (e.g. 7.0.69) has started using systemd in /usr/sbin/tomcat, which makes it not create PID-file, and ignore all the parameters used in the resource agent.

**Comment by jnpkrn** [comment]
Shouldn't it instruct the user (through, say, a warning message) to
consider migration to native systemd:tomcat agent, even if at the cost
of syncing configuration by other means?  Correct me if I'm wrong but
the purpose of pacemaker supporting systemd integration natively (as
opposed to some hypothetical `ocf:(pacemaker|heartbeat):systemdunit`
agent) is to do the job correctly by intimately interacting with systemd
API.

For instance, `is-active` quering may be broken under some circumstances
(such as the service not being enabled, which likely will be the case):

https://bugzilla.redhat.com/1073481

**Comment by oalbrigt** [comment]
Resource agents are more versatile, and can also solve issues if the systemd unit isnt acting 100% as expected.

**Comment by kskmori** [comment]
I have been using the current tomcat RA along with the upstream version of Tomcat both 7.0.* and 8.0.* without issues, which uses `catalina.sh` to invoke Tomcat. It does not seem to me that the PID file handling has changed among those version so much, so I guess that the changes you have mentioned is happened in the RHEL bundled version (point me if I'm wrong).

I am not opposed to support invocation via systemd, but users should still be able to choose the upstream version and `catalina.sh` invocation with the backward compatibility even on a systemd-enabled platform.

**Comment by jnpkrn** [comment]
There are two separate issues:
* RAs becoming just enhanced wrappers on top of systemd invocations
  - this seems like better solved with a special, hypothetical "systemd+hooks" class of agents in pacemaker itself as pacemaker is already fluent with systemd well enough
  - this is what I meant ("you are going to use systemd under the hood so why not jump on that bandvagon and use systemd unit natively")
* systemd presence does not imply preference and/or unavoidability
  - this the point raised by @kskmori

**Comment by jnpkrn** [comment]
Wider discussion of the former to emerge:
http://oss.clusterlabs.org/pipermail/developers/2016-September/000338.html

**Comment by kgaillot** [comment]
Keep in mind that pacemaker requires OCF agents to report success only after an action is completed (e.g. start should not return until the service is fully active). IIUC, systemctl does not provide this. I believe it just queues the action to be executed and returns immediately.

**Comment by oalbrigt** [comment]
I've added a "force_script" parameter so the user can avoid using systemd.

**Comment by krig** [comment]
`force_script` seems backwards to me, wouldn't it be clearer to have a parameter `init_system` which can be set to `systemd`, `sysvinit` or `auto`?

**Comment by oalbrigt** [comment]
I figured force_script was easier to implement and there's no sysvinit mode available.

**Comment by krig** [comment]
I see, alright.

File: heartbeat/tomcat [hunk 1]
```diff
@@ -56,6 +56,10 @@ else
   SU=su
 fi
 
+if which systemctl > /dev/null 2>&1; then
```
Comment (dmuhamedagic) [inline heartbeat/tomcat]: systemd/systemctl could be installed and yet not used. Probably better to test for a running daemon.

Comment (oalbrigt) [inline heartbeat/tomcat]: Good point. I'll update accordingly.

File: heartbeat/tomcat [hunk 2]
```diff
@@ -56,7 +56,7 @@ else
   SU=su
 fi
 
-if ! ocf_is_true $OCF_RESKEY_force_script && which systemctl > /dev/null 2>&1; then
+if ! ocf_is_true $OCF_RESKEY_force_script && pidof systemd > /dev/null 2>&1; then
```
Comment (jnpkrn) [inline heartbeat/tomcat]: Any reason why not checking for a known PID 1?  In general, it would likely be more reliable.

Comment (oalbrigt) [inline heartbeat/tomcat]: If PID 1 isnt systemd it should return false.

Comment (jnpkrn) [inline heartbeat/tomcat]: The point is arbitrary user on the system can do:

```
echo 'int main(void){for(;;);}' |  gcc -osystemd -xc -; ./systemd
```

which the apparently an undesired interference between the unprivileged
user and the privileged process.

Comment (oalbrigt) [inline heartbeat/tomcat]: I've updated it to check if PID 1 is systemd.

**Comment by jnpkrn** [comment]
On 27/09/16 06:20 -0700, Oyvind Albrigtsen wrote:
> I've updated it to check if PID 1 is systemd.

Getting better, but any reason to invoke extra subshell with `(list)`
construct where `{ list; }` would be just fine?  Also typical pattern
for "silent grep" used elsewhere in RAs is `grep -q` rather than
redirecting outputs to /dev/null.  And perhaps I'd also invoke
ps as `ps -o comm ...` to drop unnecessary info from output.

-- 
Jan (Poki)

**Comment by oalbrigt** [comment]
I had to group the command with ()'s to avoid getting the wrong return value in the if.

**Comment by kskmori** [comment]
The current patch will change the default behavior and will break it if you've been using the upstream version of tomcat (`catalina.sh` invocations) on a systemd-enabled platform. 

Changing the default value to opposite (force_script_default=1) could be ok, but the systemd-users would need to specify the parameter by contraries.

Can't it be more smart to decide startup methods so that it would preserve the default behavior, for example, use `catalina.sh` if it presented, and if not, use systemd when it enabled, and so on?

The tomcat RA already has some logic to detect startup scripts depending on the distros around L620-L630 (introduced by 9fc513ae through d0f63f55) and it seems to me that supporting systemd invocations is just another variant of this and it is appropriate to extend the logic here.

```
if [ -z "${TOMCAT_START_SCRIPT}" ]; then
	if [ -e "$CATALINA_HOME/bin/catalina.sh" ]; then
		TOMCAT_START_SCRIPT="$CATALINA_HOME/bin/catalina.sh"
	elif [ -e "/usr/sbin/tomcat" ]; then
		REDIRECT_DEFAULT_CONFIG=1
		TOMCAT_START_SCRIPT="/usr/sbin/tomcat"
	elif [ -e "/usr/sbin/tomcat6" ]; then
		REDIRECT_DEFAULT_CONFIG=1
		TOMCAT_START_SCRIPT="/usr/sbin/tomcat6"
	fi
fi
```

**Pull request merged**